packages/scm/src/browser/scm-service.ts

```diff
@@ -0,0 +1,325 @@
+    id: string;
+    text: string;
+    tooltip?: string;
+    command?: string;
+}
+
+export interface InputValidation {
+    message: string;
+    type: 'info' | 'success' | 'warning' | 'error';
+}
+
+export interface InputValidator {
+    (value: string): Promise<InputValidation | undefined>;
+}
+
+export namespace InputValidator {
+    /**
+     * Type for the validation result with a status and a corresponding message.
+     */
+    export type Result = Readonly<{ message: string, type: 'info' | 'success' | 'warning' | 'error' }>;
+
+    export namespace Result {
+
+        /**
+         * `true` if the `message` and the `status` properties are the same on both `left` and `right`. Or both arguments are `undefined`. Otherwise, `false`.
+         */
+        export function equal(left: Result | undefined, right: Result | undefined): boolean {
+            if (left && right) {
+                return left.message === right.message && left.type === right.type;
+            }
+            return left === right;
+        }
+
+    }
+}
+
+export interface ScmInput {
+    value: string;
+    readonly onDidChange: Event<string>;
+
+    placeholder: string;
+    readonly onDidChangePlaceholder: Event<string>;
+
+    validateInput: InputValidator;
+    readonly onDidChangeValidateInput: Event<void>;
+
+    visible: boolean;
+    readonly onDidChangeVisibility: Event<boolean>;
+}
+
+export interface ScmRepository extends Disposable {
+    readonly onDidFocus: Event<void>;
+    readonly selected: boolean;
+    readonly onDidChangeSelection: Event<boolean>;
+    readonly provider: ScmProvider;
+    readonly input: ScmInput;
+
+    focus(): void;
+
+    setSelected(selected: boolean): void;
+}
+
+@injectable()
+export class ScmServiceImpl implements ScmService {
+    private providerIds = new Set<string>();
+    private _repositories: ScmRepository[] = [];
+    private _selectedRepository: ScmRepository | undefined;
+
+    private disposableCollection: DisposableCollection = new DisposableCollection();
+    private onDidChangeSelectedRepositoriesEmitter = new Emitter<ScmRepository>();
+    private onDidAddProviderEmitter = new Emitter<ScmRepository>();
+    private onDidRemoveProviderEmitter = new Emitter<ScmRepository>();
+
+    readonly onDidChangeSelectedRepositories: Event<ScmRepository> = this.onDidChangeSelectedRepositoriesEmitter.event;
+
+    constructor() {
+        this.disposableCollection.push(this.onDidChangeSelectedRepositoriesEmitter);
+        this.disposableCollection.push(this.onDidAddProviderEmitter);
+        this.disposableCollection.push(this.onDidRemoveProviderEmitter);
+    }
+
+    get repositories(): ScmRepository[] {
+        return [...this._repositories];
+    }
+
+    get selectedRepository(): ScmRepository | undefined {
+        return this._selectedRepository;
+    }
+
+    set selectedRepository(repository: ScmRepository | undefined) {
+        this._selectedRepository = repository;
+        if (repository) {
```

Comment:
if `undefined` is a repository as well, events should respect it, otherwise clients cannot react on such changes, please review code for other such places